dali/kernels/imgproc/warp_gpu.h

```diff
@@ -0,0 +1,106 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_IMGPROC_WARP_GPU_H_
+#define DALI_KERNELS_IMGPROC_WARP_GPU_H_
+
+#include "dali/core/common.h"
+#include "dali/core/geom/vec.h"
+#include "dali/kernels/kernel.h"
+#include "dali/kernels/imgproc/warp/warp_setup.cuh"
+#include "dali/kernels/imgproc/warp/warp_variable_size_impl.cuh"
+#include "dali/kernels/imgproc/warp/warp_uniform_size_impl.cuh"
+#include "dali/kernels/imgproc/warp/mapping_traits.h"
+
+namespace dali {
+namespace kernels {
+
+/// @remarks Assume HWC layout
+template <typename _Mapping, int ndim, typename _OutputType, typename _InputType,
+          typename _BorderValue>
+class WarpGPU : public warp::WarpSetup<ndim, _OutputType, _InputType> {
+ public:
+  static constexpr int spatial_ndim = ndim;
+  static constexpr int tensor_ndim = ndim + 1;
+
+  using Mapping = _Mapping;
+  using OutputType = _OutputType;
+  using InputType = _InputType;
+  using BorderValue = _BorderValue;
+  using MappingParams = warp::mapping_params_t<Mapping>;
+  static_assert(std::is_pod<MappingParams>::value, "Mapping parameters must be POD.");
+  static_assert(std::is_pod<BorderValue>::value, "BorderValue must be POD.");
+
+  using Base =  warp::WarpSetup<spatial_ndim, OutputType, InputType>;
+  using SampleDesc = typename Base::SampleDesc;
+  using BlockDesc = typename Base::BlockDesc;
+  static_assert(spatial_ndim == 2, "Not implemented for spatial_ndim != 2");
+
+  KernelRequirements Setup(KernelContext &context,
+                           const InListGPU<InputType, tensor_ndim> &in,
+                           const InTensorGPU<MappingParams, 1> &mapping,
+                           span<const TensorShape<spatial_ndim>> output_sizes,
+                           span<const DALIInterpType> interp,
+                           BorderValue border = {}) {
+    assert(in.size() == static_cast<size_t>(output_sizes.size()));
+    auto out_shapes = this->GetOutputShape(in.shape, output_sizes);
+    return Base::Setup(out_shapes);
+  }
+
+  void Run(KernelContext &context,
+           const OutListGPU<OutputType, tensor_ndim> &out,
+           const InListGPU<InputType, tensor_ndim> &in,
+           const InTensorGPU<MappingParams, 1> &mapping,
+           span<const TensorShape<spatial_ndim>> output_sizes,
+           span<const DALIInterpType> interp,
+           BorderValue border = {}) {
+    this->ValidateOutputShape(out.shape, in.shape, output_sizes);
+    this->PrepareSamples(out, in, interp);
+    SampleDesc *gpu_samples;
+    BlockDesc *gpu_blocks;
+
+    if (this->IsUniformSize()) {
+      gpu_samples = context.scratchpad->ToGPU(context.gpu.stream, this->Samples());
```

Comment:
What's the difference between `ToContiguousGPU` and `ToGPU` and why the former is not used here, but is used in non-uniform case?

Reply:
The difference is that ToContiguousGPU returns a tuple and ToGPU returns a pointer. Since we only copy one block of data here, there's no point to wrap the result into a tuple.

Reply:
Also, since ToContiguous deals with packs, tuples and stuff, it would probably take more time to compile.

Reply:
Ok